module/VuFind/src/VuFind/ILS/Driver/Folio.php

```diff
@@ -808,33 +808,32 @@ public function placeHold($holdDetails)
         } catch (Exception $e) {
             throw new ILSException('hold_date_invalid');
         }
-        // Get status to separate Holds (on checked out items) and Pages on a
-        $status = $this->getStatus($holdDetails['item_id']);
         $requestBody = [
-            'requestType' => $item->status->name == 'Available' ? 'Page' : 'Hold',
-            'requestDate' => date('c'),
+            'instanceId' => $holdDetails['id'],
             'requesterId' => $holdDetails['patron']['id'],
-            'requester' => [
-                'firstName' => $holdDetails['patron']['firstname'] ?? '',
-                'lastName' => $holdDetails['patron']['lastname'] ?? ''
-            ],
-            'itemId' => $holdDetails['item_id'],
-            'fulfilmentPreference' => 'Hold Shelf',
+            'requestDate' => date('c'),
             'requestExpirationDate' => date_format($requiredBy, 'Y-m-d'),
+            'pickupServicePointId' => $holdDetails['pickUpLocation']
         ];
         $response = $this->makeRequest(
             'POST',
-            '/request-storage/requests',
+            '/circulation/requests/instances',
```

Comment:
As mentioned on Slack, this looks like it is going to place a title-level, rather than item-level hold; VuFind expects item-level holds by default, and will request title-level holds by passing 'title' as the 'level' key of the $holdDetails array. (See https://github.com/vufind-org/vufind/blob/master/module/VuFind/src/VuFind/ILS/Driver/VoyagerRestful.php#L1936 for the Voyager driver's example). If we can support both types of hold here, that would be great, but I don't think we want to rely purely on title-level holds, since that can lead to unexpected results (especially with multi-volume sets). Let me know if you need any clarification!

Reply:
I think FOLIO supports title-level holds (https://s3.amazonaws.com/foliodocs/api/mod-circulation/p/circulation.html#circulation_requests_instances_post)??   However it requires the instance UUID ...which we were not planning to pass into VuFind.  So at this stage in the process I have the bib HRID and the item UUID which I think I want to use to place the request.
I'm not sure how to make this all configurable / flexible.  

Also, if VuFind is going to alllow item based holds, I think it does need the hold type (page or hold).   In the holdingsils/standard.phtml file where the hold links are constructed it knows the item status...so I'm wondering if it can be appended to the url (e.g. &available=$holding['availability']  .....or something like that)?

I changed my code to be item based, added fulfilmentPreference and requestType and it works.  I'm just not sure how to pull all the options (title level/item level) together.  

I'll paste my code below.  

 $requestBody = [
            'itemId' => $holdDetails['item_id'],
            'requesterId' => $holdDetails['patron']['id'],
            'requestDate' => date('c'),
            'requestExpirationDate' => date_format($requiredBy, 'Y-m-d'),
            'pickupServicePointId' => $holdDetails['pickUpLocation'],
	    'fulfilmentPreference' => 'Hold Shelf',
	    'requestType' => 'Page'  //hardcoded for now
        ];
        $response = $this->makeRequest(
            'POST',
            '/circulation/requests',
            json_encode($requestBody)
        );

thanks!

Reply:
@mis306lu, I think we just need to use getInstanceByBibId() to look up the instance ID at the top of placeHold() -- this can be used to translate HRIDs into instance IDs in your case. Perhaps we want to create a new getInstanceIdByBibId() method so that we can short-circuit the API call in situations where we already have an instance ID.

Regarding the page vs. hold thing, the Voyager driver receives a 'holdtype' value from VuFind (see https://github.com/vufind-org/vufind/blob/c23c92c734aaae8b4b288f948e21ef9d19c7d5bc/module/VuFind/src/VuFind/ILS/Driver/VoyagerRestful.php#L1934) which it uses to place the correct type of request. I can't remember off the top of my head exactly where holdtype is coming from, but we should be able to leverage that mechanism in FOLIO as well. I'd start by var_dumping the incoming array to see what details are in there... if 'holdtype' is missing by default, let me know and I'll go digging into the remaining details.

Reply:
The incoming array to placeHold looks like this:

array(5) { ["requiredBy"]=> string(10) "06-07-2020" ["pickUpLocation"]=> string(36) "c4c90014-c8c9-4ade-8f24-b5e313319f4b" ["id"]=> string(2) "43" ["item_id"]=> string(36) "91569542-0172-4081-abc8-ad50e8cf2ff8" ["patron"]=> array(7) { ["id"]=> string(36) "1dd51453-1165-45b4-a5d5-cc948483f61e" ["username"]=> string(9) "876579559" ["cat_username"]=> string(9) "5551212" ["cat_password"]=> NULL ["firstname"]=> string(8) "Jane" ["lastname"]=> string(10) "Doe" ["email"]=> string(18) "an email address" } }

Reply:
On the subject of instanceIds vs HRIDs, I got a little turned around (see slack) and this should be using `/circulation/requests` instead of `/circulation/requests/instance` OR `/requests-storage/requests`. Like Demian pointed out this can be an item level request instead of instance like I've done here. 

There's no `auto` hold type in FOLIO to fall back on, so I think we'll need to add back in that $status variable which I took out to try to pick the correct hold type to place.

Reply:
@mis306lu, I think the way to get holdtype to show up in the incoming array is to add it as a value to the HMACKeys setting in FOLIO.ini, and to include a 'holdtype' key in the array returned by getHolding() (at the same level as the is_holdable key). The idea is that, at the same time you're looking up holding status, you can determine what type of request would be appropriate, and then that value gets passed back around to the placeHold() method. The code is not crystal clear and I'm reviewing it in a hurry, so I apologize if I lead you astray -- but I think it's worth a quick experiment. If it doesn't work, please let me know and I'll dig deeper.

Reply:
I'll give it a try.  thank you!

Reply:
I ended up with this:
added 'status' to the HMACKeys...and then I was able to add this back into the placeHold function:

	    'requestType' => $holdDetails['status'] == 'Available' ? 'Page' : 'Recall',

Checked out items AND Paged items work with both request type Recall and request type Hold.  I need to look into FOLIO to see what the difference is.  Do you think this approach is okay?    thank you!!

Reply:
Thanks @mis306lu and @demiankatz I updated the PR catch up with this conversation I think, let me know if I've misinterpreted your changes. Also updated to use the `/circulation/requests` interface to place a request by item id.

Reply:
this looks good to me.  thank you!